notify/src/address/tracker.rs

```diff
@@ -0,0 +1,649 @@
+    fn insert(&mut self, index: Index) -> bool {
+        let mut result = true;
+        self.0
+            .entry(index)
+            .and_modify(|x| {
+                *x += 1;
+                result = *x == 1;
+            })
+            .or_insert(1);
+        result
+    }
+
+    fn remove(&mut self, index: Index) -> bool {
+        let mut result = false;
+        self.0.entry(index).and_modify(|x| {
+            if *x > 0 {
+                *x -= 1;
+                result = *x == 0
+            }
+        });
+        result
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+#[cfg(test)]
+#[derive(Debug, Clone)]
+#[cfg(test)]
+pub struct Counter {
+    pub index: Index,
+    pub count: RefCount,
+    pub locked: bool,
+}
+
+#[cfg(test)]
+impl Counter {
+    pub fn new(index: Index, count: RefCount) -> Self {
+        Self { index, count, locked: false }
+    }
+
+    pub fn active(&self) -> bool {
+        self.count > 0
+    }
+}
+
+#[cfg(test)]
+impl PartialEq for Counter {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index
+    }
+}
+#[cfg(test)]
+impl Eq for Counter {}
+
+#[cfg(test)]
+impl PartialOrd for Counter {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+#[cfg(test)]
+impl Ord for Counter {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.index.cmp(&other.index)
+    }
+}
+
+/// Set of `Index`
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IndexSet(HashSet<Index>);
+
+impl IndexSet {
+    pub fn new(indexes: Vec<Index>) -> Self {
+        Self(indexes.into_iter().collect())
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self(HashSet::with_capacity(capacity))
+    }
+
+    pub fn iter(&self) -> hash_set::Iter<'_, Index> {
+        self.0.iter()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.0.capacity()
+    }
+
+    pub fn drain(&mut self) -> hash_set::Drain<'_, Index> {
+        self.0.drain()
+    }
+}
+
+impl Indexer for IndexSet {
+    fn contains(&self, index: Index) -> bool {
+        self.0.contains(&index)
+    }
+
+    fn insert(&mut self, index: Index) -> bool {
+        self.0.insert(index)
+    }
+
+    fn remove(&mut self, index: Index) -> bool {
+        self.0.remove(&index)
+    }
+
+    fn len(&self) -> usize {
+        self.len()
+    }
+
+    fn is_empty(&self) -> bool {
+        self.is_empty()
+    }
+}
+
+#[derive(Debug)]
+struct Inner {
+    script_pub_keys: IndexMap<ScriptPublicKey, RefCount>,
+    max_addresses: Option<usize>,
+    empty_entries: HashSet<Index>,
+}
+
+impl Inner {
+    const fn expand_max_addresses(max_addresses: usize) -> usize {
+        ((max_addresses + 1) * 8 / 7).next_power_of_two() * 7 / 8 - 1
```

Comment:
When `--max-tracked-addresses=0`, you get this panic:

```
rusty-kaspa git:(pr/tiram88/427-1) cargo run --bin kaspad -- --utxoindex --disable-upnp --max-tracked-addresses=0                       
    Finished dev [unoptimized + debuginfo] target(s) in 0.18s
     Running `target/debug/kaspad --utxoindex --disable-upnp --max-tracked-addresses=0`
2024-03-11 22:34:47.007-06:00 [INFO ] kaspad v0.13.5
2024-03-11 22:34:47.007-06:00 [INFO ] Application directory: /Users/dev/.rusty-kaspa
2024-03-11 22:34:47.007-06:00 [INFO ] Data directory: /Users/dev/.rusty-kaspa/kaspa-mainnet/datadir
2024-03-11 22:34:47.007-06:00 [INFO ] Logs directory: /Users/dev/.rusty-kaspa/kaspa-mainnet/logs
2024-03-11 22:34:47.007-06:00 [INFO ] Utxoindex Data directory /Users/dev/.rusty-kaspa/kaspa-mainnet/datadir/utxoindex
thread 'main' panicked at notify/src/address/tracker.rs:209:9:
attempt to subtract with overflow
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Exiting...
```

Please either make the param disallow 0, or do some handling when 0 is passed.

Reply:
We might take advantage of `--max-tracked-addresses=0` to disable the max altogether, wdyt?

Reply:
Sounds good to me. Much better than a panic. Please make sure that's documented somewhere.

Reply:
Done impl and doc, together with a revised low default (forcing to explicitly ask for a large count) and an upper bound

Reply:
Use `Option<NonZeroUsize>` it has isize layout and reflect purpose of its use better